language/move-vm/types/src/values/values_impl.rs

```diff
@@ -1988,66 +1973,57 @@ impl<'a, 'b> serde::Serialize for AnnotatedValue<'a, 'b, Type, ValueImpl> {
             (Type::Vector(ty), ValueImpl::Container(r)) => {
                 let ty = &**ty;
                 match (ty, &*r.borrow()) {
-                    (Type::Vector(_), Container::General(v))
-                    | (Type::Struct(_), Container::General(v))
-                    | (Type::Address, Container::General(v)) => {
+                    (Type::U8, Container::U8(v)) => v.serialize(serializer),
+                    (Type::U64, Container::U64(v)) => v.serialize(serializer),
+                    (Type::U128, Container::U128(v)) => v.serialize(serializer),
+                    (Type::Bool, Container::Bool(v)) => v.serialize(serializer),
+
+                    (_, Container::General(v)) => {
```

Comment:
Do we really want this to match against every type that has a `Container::General`, in the original code it explicitly matched against `Type::Vector`, `Type::Struct`, and `Type::Address`?

Reply:
I thought we should not enforce extra Move invariants if serialization does not need them. I was actually quite puzzled by these additional checks when reading the code initially.

Reply:
I'd be even fine with looking only at the value. Truth is looking at the type at all is just a massive complication for runtime.
In fact I think we should probably remove all of it (definitely at some point, not sure how soon).
Also that would not change the fact that we need a serde implementation for values for other reasons

Reply:
haha for a moment you got me @dariorussi : I just tried removing all the layouts for serialization and it doesn't work: struct are fixed size, so that they don't serialize in the same way as vectors.

Reply:
Right, currently values themselves do not carry enough information to uniquely determine the data layouts. 

@dariorussi I'd say types are still the ultimate source of truth for this matter, but there is the option to make values self descriptive, although that will limit us to certain implementations so it will need to be a conscious decision. Let's revisit this at a later time.

Reply:
It's true that having 2 types of containers in ValueImpl instead of 1 (one for struct and one for vectors) would allow to simplify the serialization code quite a lot.